File: androidVariant/src/main/java/org/comixedproject/variant/android/view/server/FileItemView.kt
/*
 * Variant - A digital comic book reading application for the iPad and Android tablets.
 * Copyright (C) 2025, The ComiXed Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses>
 */

package org.comixedproject.variant.android.view.server

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ElevatedCard
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.MaterialTheme.colorScheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import org.comixedproject.variant.android.DIRECTORY_LIST
import org.comixedproject.variant.android.R
import org.comixedproject.variant.android.VariantTheme
import org.comixedproject.variant.android.view.BYTES_PER_MB
import org.comixedproject.variant.model.library.DirectoryEntry
import org.comixedproject.variant.model.state.DownloadingState

private const val TAG = "DirectoryDetailView"

@Composable
fun FileItemView(
    fileEntry: DirectoryEntry,
    comicBookFilenameList: List<String>,
    downloadingState: List<DownloadingState>,
    onDownloadFile: (String, String) -> Unit,
    modifier: Modifier = Modifier
) {
    val downloading = downloadingState
        .filter { it.path == fileEntry.path }
        .firstOrNull()


    ElevatedCard(
        colors = CardDefaults.cardColors(containerColor = colorScheme.surface),
        modifier = modifier
            .fillMaxWidth()
    ) {
        Column(
            modifier = modifier
                .padding(8.dp)
                .fillMaxWidth()
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceBetween,
                modifier = Modifier.fillMaxWidth()
            ) {
                if (downloading == null) {
                    if (comicBookFilenameList.contains(fileEntry.filename)) {
                        IconButton(
                            onClick = { },
                            enabled = downloadingState.size < 5
                        ) {
                            Icon(
                                painterResource(id = R.drawable.ic_downloaded_file),
                                contentDescription = fileEntry.title
                            )
                        }
                    } else {
                        IconButton(
                            onClick = {
                                onDownloadFile(fileEntry.path, fileEntry.filename)
                            },
                            enabled = downloadingState.size < 5
                        ) {
                            Icon(
                                painterResource(R.drawable.ic_download_file),
                                contentDescription = fileEntry.title
                            )
                        }
                    }
                } else {
                    IconButton(
                        onClick = { },
                        enabled = downloadingState.size < 5
                    ) {
                        Icon(
                            painterResource(R.drawable.ic_downloading_file),
                            contentDescription = fileEntry.title
                        )
                    }
                }

                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = "${fileEntry.title}",
                        style = MaterialTheme.typography.bodyLarge,
                        fontWeight = FontWeight.Bold,
                        textAlign = TextAlign.Left,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )

                    if (downloading == null) {
                        Text(
                            text = "${fileEntry.filename}",
                            style = MaterialTheme.typography.bodySmall,
                            fontWeight = FontWeight.Bold,
                            textAlign = TextAlign.Left,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                    } else {
                        val received = downloading.received
                        val total = downloading.total
                        val progress = when (received > 0) {
                            true -> (received.toFloat() / total.toFloat())
                            false -> 0.0
                        }
                        val fileSize = String.format("%.1f", fileEntry.fileSize / BYTES_PER_MB)
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Text("${fileSize} MB", style = MaterialTheme.typography.bodySmall)
                            LinearProgressIndicator(
                                progress = { progress.toFloat() },
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(4.dp)
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
@Preview
fun FileItemViewPreview() {
    val fileEntry = DIRECTORY_LIST.filter { !it.isDirectory }.first()
    VariantTheme {
        FileItemView(
            fileEntry,
            listOf(),
            listOf(),
            onDownloadFile = { _, _ -> })
    }
}

@Composable
@Preview
fun FileItemViewPreviewDownloaded() {
    val fileEntry = DIRECTORY_LIST.filter { !it.isDirectory }.first()
    VariantTheme {
        FileItemView(
            fileEntry,
            listOf(fileEntry.filename),
            listOf(),
            onDownloadFile = { _, _ -> })
    }
}

@Composable
@Preview
fun FileItemViewPreviewDownloading() {
    val fileEntry = DIRECTORY_LIST.filter { !it.isDirectory }.first()
    VariantTheme {
        FileItemView(
            fileEntry,
            listOf(),
            listOf(
                DownloadingState(
                    fileEntry.path, fileEntry.filename,
                    50, 100
                )
            ),
            onDownloadFile = { _, _ -> })
    }
}

File: iosVariant/iosVariant/Views/Server/FileItemView.swift
/*
 * Variant - A digital comic book reading application for the iPad and Android tablets.
 * Copyright (C) 2025, The ComiXed Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses>
 */

import SwiftUI
import shared

private let TAG = "FileItemView"

struct FileItemView: View {
    let entry: DirectoryEntry
    let comicBookFiles: [String]
    let downloadingState: [DownloadingState]

    var onDownloadFile: (String, String) -> Void

    var downloaded: Bool {
        return comicBookFiles.contains(entry.filename)
    }

    var downloading: Bool {
        return downloadingState.map { $0.path }.contains(entry.path)
    }

    var downloadProgress: Double {
        let state = downloadingState.filter { $0.path == entry.path }.first!
        if state.total > 0 {
            return Double(state.received) / Double(state.total)
        } else {
            return 0.0
        }
    }

    var fileSize: String {
        return
            "\(String(format: "%.1f", Double(entry.fileSize) / BYTES_PER_MB))"
    }

    var body: some View {
        HStack {
            if downloaded {
                Button {

                } label: {
                    Image("downloaded_file")
                        .resizable()
                        .scaledToFit()
                        .frame(width: 32)
                }
            } else if downloading {
                Button {

                } label: {
                    Image("downloading_file")
                        .resizable()
                        .scaledToFit()
                        .frame(width: 32)
                }
            } else {
                Button {
                    Log().info(
                        tag: TAG,
                        message: "Downloading file: \(entry.path)"
                    )
                    onDownloadFile(entry.path, entry.filename)
                } label: {
                    Image("download_file")
                        .resizable()
                        .scaledToFit()
                        .frame(width: 32)
                }
            }

            VStack(alignment: .leading) {
                Text(entry.title)
                    .font(.headline)

                if downloading {
                    HStack {
                        Text("\(fileSize) MB")
                        ProgressView(value: downloadProgress)
                    }
                } else {
                    Text(entry.filename)
                        .font(.subheadline)
                }
            }
        }
    }
}

#Preview("default") {
    FileItemView(
        entry: DIRECTORY_LIST.filter { $0.isDirectory == false }.first!,
        comicBookFiles: [],
        downloadingState: [],
        onDownloadFile: { _, _ in }
    )
}

#Preview("already downloaded") {
    FileItemView(
        entry: DIRECTORY_LIST.filter { $0.isDirectory == false }.first!,
        comicBookFiles: [
            DIRECTORY_LIST.filter { $0.isDirectory == false }.first!.filename
        ],
        downloadingState: [],
        onDownloadFile: { _, _ in }
    )
}

#Preview("downloading") {
    FileItemView(
        entry: DIRECTORY_LIST.filter { $0.isDirectory == false }.first!,
        comicBookFiles: [],
        downloadingState: [
            DownloadingState(
                path: DIRECTORY_LIST.filter { $0.isDirectory == false }.first!
                    .path,
                filename: DIRECTORY_LIST.filter { $0.isDirectory == false }
                    .first!
                    .path,
                received: 50,
                total: 100
            )
        ],
        onDownloadFile: { _, _ in }
    )
}
